table/sst_file_reader.cc

```diff
@@ -0,0 +1,353 @@
+#include "rocksdb/sst_file_reader.h"
+
+#include <inttypes.h>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+#include "rocksdb/env.h"
+#include "rocksdb/iterator.h"
+#include "rocksdb/status.h"
+#include "rocksdb/table_properties.h"
+#include "table/block.h"
+#include "table/block_based_table_builder.h"
+#include "table/block_based_table_factory.h"
+#include "table/format.h"
+#include "table/meta_blocks.h"
+#include "table/plain_table_factory.h"
+#include "table/table_reader.h"
+#include "util/compression.h"
+
+#include "port/port.h"
+
+namespace rocksdb {
+
+SstFileReader::SstFileReader(
+    const std::string& file_path, bool verify_checksum,
+    std::function<void(const Slice&, const Slice&, SequenceNumber,
+                       unsigned char)>
+        kv_handler,
+    std::function<void(const std::string&)> info_handler,
+    std::function<void(const std::string&)> err_handler)
+    : file_name_(file_path),
+      read_num_(0),
+      verify_checksum_(verify_checksum),
+      kv_handler_(kv_handler),
+      info_handler_(info_handler),
+      err_handler_(err_handler),
+      ioptions_(new ImmutableCFOptions(options_)),
+      internal_comparator_(new InternalKeyComparator(BytewiseComparator())) {
+  if (info_handler_) {
+    char buf[128];
+    int len = snprintf(buf, 128, "Process %s\n", file_path.c_str());
+    info_handler_(std::string(buf, len));
```

Comment:
i learned recently that snprintf returns >= 128 when truncation happens (doesn't return the number of characters written). so this'd overrun buf in case file_path is really long.